src/util/timer.rs

```diff
@@ -0,0 +1,244 @@
+    fn partial_cmp(&self, other: &TimeoutTask<T>) -> Option<Ordering> {
+        Some(if self.next_tick > other.next_tick {
+            Ordering::Less
+        } else if self.next_tick < other.next_tick {
+            Ordering::Greater
+        } else {
+            Ordering::Equal
+        })
+    }
+}
+
+impl<T> Ord for TimeoutTask<T> {
+    fn cmp(&self, other: &TimeoutTask<T>) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+pub struct Timer<T> {
+    now: Instant,
+    tick: Duration,
+    pending: BinaryHeap<TimeoutTask<T>>,
+}
+
+impl<T> Timer<T> {
+    fn align_up_instant(&self, ins: Instant) -> Instant {
+        let mut next = self.now + self.tick;
+        let now = Instant::now_coarse();
+        loop {
+            if next < ins || next < now {
+                next += self.tick;
+                continue;
+            }
+            break next;
+        }
+    }
+
+    fn deliver(&self, mut task: TimeoutTask<T>, buffer: &mut Vec<TimeoutTask<T>>) {
+        task.next_tick = self.align_up_instant(task.next_tick + task.timeout);
+        buffer.push(task);
+    }
+
+    /// Create a new timer with a specified `tick` which means
+    /// the min timeout when the `Timer` calls `next_timeout`.
+    pub fn new(tick: Duration) -> Self {
+        Timer {
+            now: Instant::now_coarse(),
+            tick: tick,
+            pending: BinaryHeap::new(),
+        }
+    }
+
+    /// Add a periodic task into the `Timer`. The `timeout` will
+    /// be aligned up to `tick` of the `Timer` if need.
+    pub fn add_task(&mut self, timeout: Duration, task: T) {
+        let task = TimeoutTask {
+            next_tick: self.now + timeout,
+            timeout: timeout,
+            task: task,
+        };
+        self.pending.push(task);
+    }
+
+    /// Get the next `timeout` from the timer, and fill `delivered`
+    /// with tasks will be triggered after `timeout` elapsed.
+    /// If there is no pending tasks, `delivered` won't be changed.
+    pub fn next_timeout(&mut self, delivered: &mut Vec<TimeoutTask<T>>) -> Option<Duration> {
+        if let Some(timeout_task) = self.pending.pop() {
+            let next_tick = self.align_up_instant(timeout_task.next_tick);
+            self.deliver(timeout_task, delivered);
+            while let Some(timeout_task) = self.pending.pop() {
+                if timeout_task.next_tick > next_tick {
+                    self.pending.push(timeout_task);
+                    break;
+                }
+                self.deliver(timeout_task, delivered);
+            }
+            self.now = next_tick;
+            // TODO: Should use checked sub.
+            return Some(next_tick - Instant::now_coarse());
+        }
+        None
+    }
+
+    pub fn as_wrapper(&mut self) -> TimerWrapper<T> {
+        TimerWrapper(self)
+    }
+}
+
+pub struct TimerWrapper<'a, T: 'a>(&'a mut Timer<T>);
+
+impl<'a, T> TimerWrapper<'a, T> {
+    pub fn restore(&mut self, task: TimeoutTask<T>) {
```

Comment:
I don't know why we need `restore` here, adding the task back again with a new timeout is enough.